packages/Manager/src/test/kotlin/nz/govt/eop/plan_limits/WaterAllocationAndUsageViewsTest.kt

```diff
@@ -0,0 +1,90 @@
+package nz.govt.eop.plan_limits
+
+import io.kotest.matchers.collections.shouldContain
+import io.kotest.matchers.shouldBe
+import java.time.LocalDate
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.springframework.beans.factory.annotation.Autowired
+import org.springframework.boot.test.context.SpringBootTest
+import org.springframework.jdbc.core.JdbcTemplate
+import org.springframework.test.context.ActiveProfiles
+import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
+import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
+import org.springframework.transaction.annotation.Transactional
+
+@ActiveProfiles("test")
+@SpringBootTest
+@Transactional
+class WaterAllocationAndUsageViewsTest(@Autowired val jdbcTemplate: JdbcTemplate) {
+
+  @BeforeEach
+  fun resetTestData() {
+    truncateTestData()
+    createTestData()
+  }
+
+  @Test
+  fun `should be empty with no data`() {
+    truncateTestData()
+    val result =
+        jdbcTemplate.queryForObject(
+            "select count(*) from water_allocation_and_usage_by_area", Int::class.java)
+    result shouldBe 0
+  }
+
+  @Test
+  fun `should include a year of data`() {
+    val result =
+        jdbcTemplate.queryForMap(
+            "select count(*) as count, min(date) as min_date, max(date) as max_date from water_allocation_and_usage_by_area")
+    arrayOf(365L, 366L) shouldContain result["count"] as Long
+    val min = LocalDate.parse(result["min_date"].toString())
+    val max = LocalDate.parse(result["max_date"].toString())
+    min.plusYears(1) shouldBe max
+  }
+
+  // Pending tests
+  // Single consent
+  fun `should not include data before a consent was effective`() {}
+  fun `should data once a consent is active`() {}
+  fun `should handle changes in consent data`() {}
+  fun `should not include data when a consent status is not active`() {}
+  fun `should not include observations when is_metered is false`() {}
+
+  // Multiple consents
+  // Multiple meters
```

Comment:
@smozely I've started writing some tests for the view in this style. Feels like they could quickly get verbose without some helpers for generating sets of test data. This test data needs to be dynamic, since the new views generate data relative to now.

Any thoughts on things in the Java/Kotlin/Spring ecosystem to help with this?

Reply:
Ah ... not sure I have any great answers

We had the same kinda thing when trying to test complex views at ECAN, we had written a small JS layer to do similar to this style of testing, we did struggle with how to express in the code what the test data means as it was difficult to jump from "these rows here, mean this output in the view" ... not sure any suggestion here really helps with that.

via DBT we did have an a way of testing constituent views in stages, well actually it was less than perfect in that we could only test the views in stages, not the whole chain together.


Keeping it kinda how you've got there Kotlin lets you add parameters with defaults, and you can pass named parameters. Which would kinda give you the equivalent of passing an object in JS and then it being merged with a default object.

```
  fun createTestData(
      measurementName: String = "Water Meter Reading",
      consentAreaId: String = "area-id",
      consentSourceId: String = "source-id"
  ) {

/// STUFF
```

Which can then be called like 
```
createTestData(consentSourceId = "foo", measurementName = "bar")
```
I think that is a good way to keep the same core test data setup but let you customize it per test 

And then for the dates stuff, using the date `LocalDate.now().atStartOfDay().minusDays(10)` to make everything relative to now ... its kinda ugly because its hard to keep that date mangling noise down. 

- Maybe there could be the equivalent of `createTestData` something like `createExpectedResult` that would help encapsulate the logic in code, in a way that you can then use `SELECT * FROM water_allocation_and_usage_by_area` and compare directly to that? ... it might mean each test is doing more work, but keeps some consistency of how they flow

- Some of this would be a bit easier if we had objects that were being returned from the view queries, just because it would be easier to create those than the `Map` rows to assert against. But that might be overkill if they are only going to be used for tests.